cmd/bashbrew/docker.go

```diff
@@ -237,16 +242,65 @@ func (r Repo) dockerBuildUniqueBits(entry *manifest.Manifest2822Entry) ([]string
 	return uniqueBits, nil
 }
 
-func dockerBuild(tag string, file string, context io.Reader, extraEnv []string) error {
+func dockerBuild(tag string, file string, context io.Reader, platform string) error {
 	args := []string{"build", "--tag", tag, "--file", file, "--rm", "--force-rm"}
 	args = append(args, "-")
 	cmd := exec.Command("docker", args...)
-	if extraEnv != nil {
-		cmd.Env = append(os.Environ(), extraEnv...)
+	cmd.Env = append(os.Environ(), "DOCKER_BUILDKIT=0")
+	if debugFlag {
+		fmt.Println("$ export DOCKER_BUILDKIT=0")
+	}
+	if platform != "" {
+		// ideally, we would set this via an explicit "--platform" flag on "docker build", but it's not supported without buildkit until 20.10+ and this is a trivial way to get Docker to do the right thing in both cases without explicitly trying to detect whether we're on 20.10+
+		// https://github.com/docker/cli/blob/v20.10.7/cli/command/image/build.go#L163
+		cmd.Env = append(cmd.Env, "DOCKER_DEFAULT_PLATFORM="+platform)
 		if debugFlag {
-			fmt.Printf("$ export %q\n", extraEnv)
+			fmt.Printf("$ export DOCKER_DEFAULT_PLATFORM=%q\n", platform)
+		}
+	}
+	cmd.Stdin = context
+	if debugFlag {
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		fmt.Printf("$ docker %q\n", args)
+		return cmd.Run()
+	} else {
+		buf := &bytes.Buffer{}
+		cmd.Stdout = buf
+		cmd.Stderr = buf
+		err := cmd.Run()
+		if err != nil {
+			err = cli.NewMultiError(err, fmt.Errorf(`docker %q output:%s`, args, "\n"+buf.String()))
 		}
+		return err
+	}
+}
+
+const (
+	dockerfileSyntaxKey     = "BASHBREW_BUILDKIT_SYNTAX"
+	dockerfileSyntaxDefault = "docker/dockerfile:1"
+)
+
+func dockerBuildx(tag string, file string, context io.Reader, platform string) error {
+	dockerfileSyntax, ok := os.LookupEnv(dockerfileSyntaxKey)
+	if !ok {
+		dockerfileSyntax = dockerfileSyntaxDefault
+	}
+
+	args := []string{
+		"buildx",
+		"build",
+		"--progress", "plain",
+		"--build-arg", "BUILDKIT_SYNTAX=" + dockerfileSyntax,
+		"--tag", tag,
+		"--file", file,
 	}
+	if platform != "" {
+		args = append(args, "--platform", platform)
```

Comment:
Hmm, one of the reasons we _don't_ set this on the classic build is because it then also affects `FROM` instructions, and might make the engine complain about the architecture of the image that we already carefully pulled the correct version of, or worse, actually pull a _different_ image because it doesn't think the one we have is correct (basically, the fact that this implies a platform _vector_ on pulling necessary images is a huge problem for reliably building images :grimacing:).

I think we probably need the same `fromScratch` condition on setting this so that it only applies the correct metadata on the bottom layer of all our builds and doesn't do so elsewhere?  Does `buildx`/BuildKit do anything "smart" here that we might need to be careful of, like applying the current platform if one isn't specified?  Is there a way to specify "this platform and only this platform exactly" ?  (I'm not aware of anywhere in either Docker _or_ containerd's usage of these platform strings that has precedence for this. :see_no_evil:)

Reply:
Have added the logic back in :smile:

Reply:
> like applying the current platform if one isn't specified

Buildkit pulls the image in for the metadata, but I don't think it does it affects what would be pulled.